plugin/pkg/admission/initialresources/hawkular.go

```diff
@@ -17,18 +17,195 @@ limitations under the License.
 package initialresources
 
 import (
+	"crypto/tls"
+	"crypto/x509"
 	"fmt"
+	"github.com/golang/glog"
+	"github.com/hawkular/hawkular-client-go/metrics"
+	"io/ioutil"
+	"k8s.io/kubernetes/pkg/api"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
 	"time"
 
-	"k8s.io/kubernetes/pkg/api"
+	kube_client "k8s.io/kubernetes/pkg/client/unversioned"
+	kubeClientCmd "k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
 )
 
-type hawkularSource struct{}
+type hawkularSource struct {
+	client       *metrics.Client
+	uri          *url.URL
+	useNamespace bool
+	modifiers    []metrics.Modifier
+}
+
+const (
+	containerImageTag string = "container_base_image"
+	descriptorTag     string = "descriptor_name"
+	separator         string = "/"
+
+	defaultServiceAccountFile = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+)
 
-func newHawkularSource() (dataSource, error) {
-	return nil, fmt.Errorf("hawkular source not implemented")
+// Get equivalent MetricDescriptor.Name used in the Heapster
+func heapsterEqName(kind api.ResourceName) string {
+	switch kind {
+	case api.ResourceCPU:
+		return "cpu/usage"
+	case api.ResourceMemory:
+		return "memory/usage"
+	default:
+		return ""
+	}
 }
 
-func (s *hawkularSource) GetUsagePercentile(kind api.ResourceName, perc int64, image, namespace string, exactMatch bool, start, end time.Time) (int64, int64, error) {
-	return 0, 0, fmt.Errorf("gcm source not implemented")
+// Create tagFilter query
+func tagQuery(kind api.ResourceName, image string, exactMatch bool) map[string]string {
+	q := make(map[string]string)
+
+	// Add here the descriptor_tag..
+	q[descriptorTag] = heapsterEqName(kind)
+
+	if exactMatch {
+		q[containerImageTag] = image
+	} else {
+		q[containerImageTag] = fmt.Sprintf("%s:*", strings.Split(image, ":")[0])
+	}
+
+	return q
+}
+
+// dataSource API
+
+func (self *hawkularSource) GetUsagePercentile(kind api.ResourceName, perc int64, image, namespace string, exactMatch bool, start, end time.Time) (int64, int64, error) {
```

Comment:
Most Go code tends not to use `self`. You might want to change it to something like `h` or `hs` or `source`.